### ISSUE TITLE
apiserver stops working after enable plugin

### ISSUE THREAD

**Issue by sandroden** [canonical/microk8s#2586]
I have several different crashes in a 2 nodes cluster I'm using to study. 

I write 'crashes' but what happens seems it's the API server that won't come up.

I had the first after joining the second node. At that point I rebooted the machine and it started again. Now I just enabled dashboard and prometheus. Microk8s announced it was rebooting

```
Configuring node 213.215.x.y
Configuring node 212.239.z.w
Restarting nodes.
The connection to the server 127.0.0.1:16443 was refused - did you specify the right host or port?
```
If I run `microk8s.daemon-apiserver` , I get `Will not run along with kubelite`
```
sudo /usr/bin/snap run microk8s.daemon-apiserver
+ export PATH=/snap/microk8s/2407/usr/sbin:/snap/microk8s/2407/usr/bin:/snap/microk8s/2407/sbin:/snap/microk8s/2407/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
+ PATH=/snap/microk8s/2407/usr/sbin:/snap/microk8s/2407/usr/bin:/snap/microk8s/2407/sbin:/snap/microk8s/2407/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
++ /snap/microk8s/2407/bin/uname -m
+ ARCH=x86_64
+ export LD_LIBRARY_PATH=:/snap/microk8s/2407/lib:/snap/microk8s/2407/usr/lib:/snap/microk8s/2407/lib/x86_64-linux-gnu:/snap/microk8s/2407/usr/lib/x86_64-linux-gnu
+ LD_LIBRARY_PATH=:/snap/microk8s/2407/lib:/snap/microk8s/2407/usr/lib:/snap/microk8s/2407/lib/x86_64-linux-gnu:/snap/microk8s/2407/usr/lib/x86_64-linux-gnu
+ export LD_LIBRARY_PATH=/var/lib/snapd/lib/gl:/var/lib/snapd/lib/gl32:/var/lib/snapd/void::/snap/microk8s/2407/lib:/snap/microk8s/2407/usr/lib:/snap/microk8s/2407/lib/x86_64-linux-gnu:/snap/microk8s/2407/usr/lib/x86_64-linux-gnu
+ LD_LIBRARY_PATH=/var/lib/snapd/lib/gl:/var/lib/snapd/lib/gl32:/var/lib/snapd/void::/snap/microk8s/2407/lib:/snap/microk8s/2407/usr/lib:/snap/microk8s/2407/lib/x86_64-linux-gnu:/snap/microk8s/2407/usr/lib/x86_64-linux-gnu
+ '[' -e /var/snap/microk8s/2407/var/lock/lite.lock ']'
+ echo 'Will not run along with kubelite'
Will not run along with kubelite
+ exit 0

```

In case it is relevant. The nodes have 2 interfaces each: 1 public and one private and I joined them via the private one.

[inspection-report-20210922_154602.tar.gz](https://github.com/ubuntu/microk8s/files/7212130/inspection-report-20210922_154602.tar.gz)

EDIT: rebooting the node restored the cluster.
microk8s  v1.21.4   2407   1.21/stable      canonical✓  classic

**Comment by ktsakalozos** [comment]
Hi @sandroden 

I see in the logs this error message:
```
Sep 22 15:08:21 adk-c1 microk8s.daemon-kubelite[2003538]: Error: create node: failed to set bind address "@snap.microk8s.dqlite-3297041220608546238": 1
Sep 22 15:08:21 adk-c1 microk8s.daemon-kubelite[2003538]: F0922 15:08:21.444647 2003538 daemon.go:67] API Server exited create node: failed to set bind address "@snap.microk8s.dqlite-3297041220608546238": 1
```

I think we have seen this kind of message in the past in deployments with glusterfs. @balchua may have more insight on this.

**Comment by balchua** [comment]
We've never found the reason why gluster seems to be holding that sock file. ☹️

**Comment by sandroden** [comment]
Thanks @ktsakalozos ,

I confirm I'm using glusterfs...

Is there a workaround? does that mean I'll experience this any time I add a node or a plugin? Is there anything I can do to help understand it?

**Comment by balchua** [comment]
Based on previous diagnosis this seem to happen randomly.  So far it resolves itself after a reboot as some users claim. is gluster running as pods?

**Comment by sandroden** [comment]
No, I'm using glusterfs *client*, not server. The server is a dedicated server and not managed by me.
The server I'm using is Ubuntu 20.04 with glusterfs9 prepackaged .deb.

**Comment by balchua** [comment]
@sandroden is there any logs in gluster that can help us trace this down?
My knowledge of gluster is non existent tho. ☹️

**Comment by tijs2** [comment]
We have the same problem and we also are only using glusterfs as client.

What I did see is that when I do a fresh install everything is working until I reboot the system when pods are using glusterfs.

So could it be the startup order, because the k8s/dqlite system is running before I am booting the first pods but on a reboot I see that it starts pods quite quickly and maybe the k8s/dqlite system is starting later.
So if we change it (and not sure if it can be done, and also more as a quick fix) that first the kubelite dqlite process starts and then the pods?

It is only a guess and I also do not have a lot of glusterfs knowledge.

**Comment by balchua** [comment]
Thanks @tijs2
I was just wondering how the pod starts when kubelite/dqlite is not ready.
This is a good lead.

**Comment by tijs2** [comment]
I was doing a bit more of experimenting and I disabled the containerd service, killed all glusterfs things
Then restarted kubelet but kubelet is also depending on containerd so I then started the containerd service again.
I then did get access to the api with `kubectl get pods` for a few seconds until the socket is taken over by glusterfs :(

`netstat -anp` indeed gives glusterfs as the process owner of the socket which a few seconds ago it was not:
from
unix  2      [ ACC ]     STREAM     LISTENING     243632   7630/kubelite        @snap.microk8s.dqlite-3297041220608546238
to
unix  2      [ ACC ]     STREAM     LISTENING     175025   29482/glusterfs      @snap.microk8s.dqlite-3297041220608546238

In short I think the starting order does not fix the problem weird enough because the socket owner changes. Why a reinstall helped no idea but it actually did because I would expect it to behave the same.

**Comment by balchua** [comment]
Thanks @tijs2 this is helpful.
Somewhere the gluster controller in kubernetes perhaps is starting the gluster client.

Is there some gluster client logs that we can inspect?

**Comment by tijs2** [comment]
Hi @balchua 

I have no logs at least that I could find, the /var/log/glusterfs folder is empty.

But what I was thinking if it happens with containerd does that mean we should open a ticket there as well or could it be something else? I must admit that I do not know. Which process actually mounts / runs the glusterfs commands?

What I did check is that I am using 1.21/stable from microk8s and that the version of containerd is "github.com/containerd/containerd v1.4.4 05f951a3781f4f2c1911b05e61c160e9c30eaa8e"

There are newer version available and one of them mentions this: "Fix leaking socket path in runc shim v2", if this is even remotely related no idea but I was triggered by the word socket: https://github.com/containerd/containerd/releases/tag/v1.4.5

On a side note I read that there are a few security fixes on newer versions of containerd so upgrading seems like a good idea anyway

**Comment by balchua** [comment]
I can't tell for sure if this is an issue with containerd or gluster.
I have almost zero knowledge on gluster tbh.
Do u mind posting the sample yaml file where you mount your gluster persistent volume?

Maybe i should read more on gluster hahaha 😂

**Comment by tijs2** [comment]
Haha I also tried to find something on the gluster page with this issue in mind but there where no similar tickets that I could find.

Is this what you mean:
apiVersion: v1
kind: PersistentVolume
metadata:
  name: storage
  annotations:
    pv.beta.kubernetes.io/gid: "590"
spec:
  storageClassName: storage
  capacity:
    storage: 100Gi
  accessModes:
    - ReadWriteOnce
  glusterfs:
    endpoints: glusterfs-cluster
    endpointsNamespace: storage
    path: /data0/storage
    readOnly: false
  persistentVolumeReclaimPolicy: Retain
---
apiVersion: v1
kind: Endpoints
metadata:
  name: glusterfs-cluster
subsets:
- addresses:
  - ip: x.x.x.x
  - ip: y.y.y.y
  - ip: z.z.z.z
  ports:
  - port: 1
---
apiVersion: v1
kind: Service
metadata:
  name: glusterfs-cluster
spec:
  ports:
  - port: 1

I did remove the ip addresses but this is what I do to create the volume claim.

**Comment by balchua** [comment]
I see that this is the in-tree based gluster persistent volume provider.  I think this is something which will eventually go away, in favor of CSI.

Looking at current documentation, it seems like they promote using heketi and gluster.
But my brain is still confused. Haha.

**Comment by tijs2** [comment]
The weird thing is is that heketi is no longer actively supported https://github.com/heketi/heketi so feels a bit weird to start using it.

And the old way seems to be supported: https://kubernetes.io/docs/concepts/storage/volumes/#glusterfs at least it does not say it is deprecated

**Comment by balchua** [comment]
@tijs2 you are absolutely right. Hehehe

**Comment by tijs2** [comment]
I found something else what maybe could help.
What I did was
- reinstall microk8s
- Install the pods with the glusterfs mounts
- Monitored and see if everything kept running which it did.
- stopped microk8s

Then I found out that the glusterfs mounts are still alive and only then took over the socket, which was not the case when microk8s was running since the fresh install.

That raises 3 questions for me:
- Why are the glusterfs mounts not stopped when microk8s is stopped
- Is there some kind of inheritance that if the master program is stopped the child, which in this case is glusterfs, inherits the socket and keeps it open? If I look at the pid of gluster there is no parent pid so maybe that is not the case?
- What is the difference between install and then adding applications and a restart.

**Comment by balchua** [comment]
My educated guess is that when kubelite process dies the unix socket is handed over to its child process.  To test this theory, in the master branch microk8s now start dqlite into its own process instead of being part of the apiserver / kubelite.

I think this version is in the `latest/edge` channel.
If our theory is right your question #2 may hold the answer.

@ktsakalozos is the `latest/edge` already have the dqlite split?

**Comment by tijs2** [comment]
Hi @balchua and @ktsakalozos,

I was wondering if there is any progress on this ticket because our production systems are affected by this issue.

Tijs

**Comment by balchua** [comment]
@tijs2 did u get the chance to try the `latest/edge` channel?  If our theory is right, then the issue may not show up.

**Comment by tijs2** [comment]
Sorry did not know that I could test it already :)
Will try it out when I find the time to install it.

**Comment by tijs2** [comment]
It seems to solve the problem, I can now reboot the machine and everything will come up again without issues

**Comment by balchua** [comment]
Aaa thanks @tijs2 for nailing this one.

Unfortunately the feature to separate dqlite from kubelite might only be available on the next kubernetes release.

**Comment by tijs2** [comment]
Do you when it will be released? Or is there no estimation yet?

**Comment by balchua** [comment]
According to this https://www.kubernetes.dev/resources/release/
The date is on 7 december 2021.  MicroK8s usually ships about the same time or a day or two later.

**Comment by balchua** [comment]
I think we are ready to close this issue.  Feel free to re-open when its not. Thanks
